<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Instructions" data-toc-modified-id="Instructions-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Instructions</a></span></li><li><span><a href="#Before-You-Begin" data-toc-modified-id="Before-You-Begin-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Before You Begin</a></span></li><li><span><a href="#Imports" data-toc-modified-id="Imports-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Imports</a></span></li><li><span><a href="#Questions" data-toc-modified-id="Questions-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Questions</a></span><ul class="toc-item"><li><span><a href="#Optimal-Bond-Allocation-for-the-USC-Investment-Office" data-toc-modified-id="Optimal-Bond-Allocation-for-the-USC-Investment-Office-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Optimal Bond Allocation for the USC Investment Office</a></span><ul class="toc-item"><li><span><a href="#Decision-Variables" data-toc-modified-id="Decision-Variables-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Decision Variables</a></span></li><li><span><a href="#Objective" data-toc-modified-id="Objective-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>Objective</a></span></li><li><span><a href="#Constraints" data-toc-modified-id="Constraints-4.1.3"><span class="toc-item-num">4.1.3&nbsp;&nbsp;</span>Constraints</a></span></li><li><span><a href="#Solution" data-toc-modified-id="Solution-4.1.4"><span class="toc-item-num">4.1.4&nbsp;&nbsp;</span>Solution</a></span></li></ul></li><li><span><a href="#Bond-Dedication" data-toc-modified-id="Bond-Dedication-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Bond Dedication</a></span><ul class="toc-item"><li><span><a href="#Decision-Variables" data-toc-modified-id="Decision-Variables-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Decision Variables</a></span></li><li><span><a href="#Objective" data-toc-modified-id="Objective-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Objective</a></span></li><li><span><a href="#Constraints" data-toc-modified-id="Constraints-4.2.3"><span class="toc-item-num">4.2.3&nbsp;&nbsp;</span>Constraints</a></span></li><li><span><a href="#Solution" data-toc-modified-id="Solution-4.2.4"><span class="toc-item-num">4.2.4&nbsp;&nbsp;</span>Solution</a></span></li><li><span><a href="#ASIDE:-Abstract-(General)-Formulation---Linear-Optimization-Model-for-Bond-Dedication" data-toc-modified-id="ASIDE:-Abstract-(General)-Formulation---Linear-Optimization-Model-for-Bond-Dedication-4.2.5"><span class="toc-item-num">4.2.5&nbsp;&nbsp;</span>ASIDE: Abstract (General) Formulation - Linear Optimization Model for Bond Dedication</a></span></li><li><span><a href="#ASIDE:-Some-Practical-Details-About-Bonds" data-toc-modified-id="ASIDE:-Some-Practical-Details-About-Bonds-4.2.6"><span class="toc-item-num">4.2.6&nbsp;&nbsp;</span>ASIDE: Some Practical Details About Bonds</a></span></li></ul></li><li><span><a href="#Quadratic-Asset-Allocation" data-toc-modified-id="Quadratic-Asset-Allocation-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Quadratic Asset Allocation</a></span><ul class="toc-item"><li><span><a href="#Decision-Variables" data-toc-modified-id="Decision-Variables-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Decision Variables</a></span></li><li><span><a href="#Objective" data-toc-modified-id="Objective-4.3.2"><span class="toc-item-num">4.3.2&nbsp;&nbsp;</span>Objective</a></span></li><li><span><a href="#Constraints" data-toc-modified-id="Constraints-4.3.3"><span class="toc-item-num">4.3.3&nbsp;&nbsp;</span>Constraints</a></span></li><li><span><a href="#Solution" data-toc-modified-id="Solution-4.3.4"><span class="toc-item-num">4.3.4&nbsp;&nbsp;</span>Solution</a></span></li></ul></li></ul></li></ul></div>

<h1>Instructions</h1>
<br>
<font size="+1">
    <ul>
        <li style="color:red">This problem set assesses your ability to analyze a given business decision by creating a concrete formulation of an optimization model, as well as practice solving your optimizaton model with CVXPY and summarizing the results. You should typeset math within Markdown using $\LaTeX$.</li>
        <br>
        <li><b>Recommended Steps</b></li>
        <br>
        <ol>
            <li>Read the prompt and summarize the task in your own words, using as few words as possible for ease of comprehension.</li>
            <br>
            <li>Pretend to be the computer program that you are trying to create. Examine the smaple inputs and manually produce the appropriate results. Afterward, verify the correctness of your logic by comparing with the given sample outputs.</li>
            <br>
            <li>On a scrap piece of paper, write step-by-step instructions so that someone else can pretend to be the computer program and produce the appropriate results for any possible inputs. These instructions should target a human audience who does not know the problem description and only has your instructions as a guide. These instructions should serve as the roadmap for your code.</li>
            <br>
            <li>On another scrap piece of paper, write fragments of Python code to implement each step in your roadmap, focusing especially on the trickiest parts of the logic.</li>
            <br>
            <li>With your roadmap and code fragments from the previous steps, type Python code into a new Jupyter notebook cel to solve the given problem. You should type incrementally and run the code whenever you add a small chunk. This allows you to correct errors as they occur. It may also be helpful to print intermediate results to verify that the code is carrying out the logic as desired.</li>
            <br>
            <li>Run your code with some sample inputs and verify that the outputs are exactly the same as the sample outputs.</li>
            <br>
            <li>Once all of your problems are finished, it is useful to click <i>Kernel</i> $\rightarrow$ <i>Restart & Run All</i> to verify that your outputs for all of your problems continue to be correct. This helps to ensure that someone else will be able to replicate your results when running your notebook.</li>
            <br>
            <li style="color:red">Submit the <i>.ipynb</i> file on TurnItIn on Blackboard by the deadline. Be sure to refresh the page and double check your submission actually went through. <b>Note that you only need to submit your solutions, not all of the other recommended steps.</b> The recommended steps are meant to serve as a guide for your thinking process.</li>
            <br>
        </ol>
        <br>
        <hr style="border: 10px solid black">
        <br>
        <li><b>Grading</b></li>
        <br>
        <li>There are four possible scores you can get from submitting this assignment on time (submitting a blank file or one without any apparent effort does not count). Note that the rubric is designed to incentivize you to go for $100%$ mastery of the material, as the little details matter in programming.</li>
        <br>
        <ul>
            <li>Grade of $5$ out of $5$ - perfect submission with no significant errors</li>
            <br>
            <li>Grade of $4$ out of $5$ - near perfect submission with one or more significant errors</li>
            <br>
            <li>Grade of $2$ out of $5$ - apparent effort but far from perfect</li>
            <br>
            <li>Grade of $0$ out of $5$ - no submission or no apparent effort</li>
            <br>
        </ul>
    </ul>
</font>

<hr style="border: 20px solid black">

<h1>Before You Begin</h1>
<br>
<font size="+1">
    <ul>
        <li>Please read: <b>01_Portfolio_Optimization.ipynb</b></li>
        <br>
        <li>Please read: <b>02_Indexing_and_the_Rise_of_ETFs.ipynb</b></li>
        <br>
    </ul>
</font>

<hr style="border: 20px solid black">

<h1>Imports</h1>
<br>
<font size="+1">
    <ul>
        <li>Put your imports below.</li>
        <br>
        <li>Import Numpy and Pandas and CVXPY</li>
        <br>
    </ul>
</font>

In [1]:
import numpy as np
import pandas as pd
import cvxpy as cp

<hr style="border: 20px solid black">

<h1>Questions</h1>

<h2>Optimal Bond Allocation for the USC Investment Office</h2>
<br>
<font size="+1">
    <ul>
        <li>Assume the <a href="http://departmentsdirectory.usc.edu/investment_office.html">USC Investment Office</a> has decided they want to allocate $\$100,000$ of the endowment fund into two different bonds: a corporate bond, and a government bond. The bonds have the following</li>
        <br>
        <ul>
        <li><a href="https://www.investopedia.com/terms/b/bond-yield.asp">yield</a> (coupon relative to current price),</li>
        <br>
        <li>risk level (some KPI statistic that measures risk),</li>
        <br>
        <li>and <a href="https://en.wikipedia.org/wiki/Bond_(finance)#Maturity">maturity</a> (time before your principal is returned).</li>
        <br>
        </ul>
        <li><table>
  <tr>
    <th>Bond</th>
    <th>Yield</th>
    <th>Risk Level</th>
    <th>Maturity</th>
  </tr>
  <tr>
    <td>Corporate</td>
    <td>$4\%$</td>
    <td>2</td>
    <td>3 years</td>
  </tr>
  <tr>
    <td>Government</td>
    <td>$3\%$</td>
    <td>1</td>
    <td>4 years</td>
  </tr>
</table></li>
        <br>
        <li>One of the endowment's investment officer would like allocate the $\$100,000$ slice of the endowment so that</li>
        <br>
        <ul>
            <li>the <b>average</b> risk level of the portfolio is at most $1.5$,</li>
            <br>
            <li>the <b>average</b> maturity is at most $3.6$ years,</li>
            <br>
            <li>and any amount not invested in the bonds will be kept in a cash account that is assumed to generate no interest and does not contribute to the average risk level; that is, assume cash has zero yield, zero risk level, and zero maturity.</li>
            <br>
        </ul>
        <li style="color:blue">The goal of this exercise is to formulate and solve a linear optimization problem.</li>
        <br>
        <li style="color:red">NOT REQUIRED: How might you write these objectives and constraints in matrix notation?</li>
        <br>
    </ul>
</font>

<h3>Decision Variables</h3>
<br>
<font size="+1">
    <ul>
        <li>What are the decision variables and how should you formulate them?</li>
        <br>
    </ul>
</font>

In [2]:
portfolio_weights = cp.Variable(3, nonneg=True)

weight_Corp = portfolio_weights[0]

weight_Gov = portfolio_weights[1]

weight_Cash = portfolio_weights[2]

<h3>Objective</h3>
<br>
<font size="+1">
    <ul>
        <li>What is the objective USC's investment office seeks to optimize (maximize or minimize) and how should they formulate the objective using the decision variables and the parameters about the bonds?</li>
        <br>
    </ul>
</font>

- We want to maximize the expected return while controlling that the risk level of the entire portfolio is no higher than 1.5

In [3]:
expected_return = weight_Corp *0.04 + weight_Gov * 0.03 + weight_Cash * 0

objective = cp.Maximize(expected_return)

<h3>Constraints</h3>
<br>
<font size="+1">
    <ul>
        <li>What are the constraints imposed by USC's investment office, and how should they be formulated using the decision variables and the parameters about the bonds?</li>
        <br>
        <ul>
            <li><i>Hint: the average risk level is equal to $\frac{\text{\$ of risk}}{\text{\$ total}}$, and $\text{\$ of risk}$ equals the sum of the products of the risk level for a given bond and the dollars allocated to a given bond</i></li>
            <br>
            <li><i>Hint: the average maturity is equal to $\frac{\text{\$ in maturity}}{\text{\$ total}}$, and $\text{\$ in maturity}$ equals the sum of the products of the maturity for a given bond and the dollars allocated to a given bond</i></li>
            <br>
        </ul>
    </ul>
</font>

In [4]:
avg_risk_level = (weight_Corp * 100_000 * 2 + weight_Gov *100_000 * 1) / 100_000

avg_maturity = (weight_Corp * 100_000 * 3 + weight_Gov *100_000 * 4) / 100_000

In [5]:
constraint_1 = (avg_risk_level <=1.5)

constraint_2 = (avg_maturity <= 3.6)

constraint_3 = (weight_Corp + weight_Gov + weight_Cash ==1)

constraints = [constraint_1, constraint_2, constraint_3]

<h3>Solution</h3>
<br>
<font size="+1">
    <ul>
        <li>What is the solution to the problem?</li>
        <br>
        <li>How can you interpret the solution?</li>
        <br>
    </ul>
</font>

In [6]:
prob = cp.Problem(objective, constraints)

In [7]:
prob.solve()

0.034999995901758424

In [8]:
np.round(portfolio_weights.value,2)

array([0.5, 0.5, 0. ])

- Intepretation: the USC Investment Office should invest 50% of their money in corporate bonds, and 50% in government bonds, which will yield an expected return of 3.5%.

<br>
<font size="+1">
    <ul>
        <li style="color:red">NOT REQUIRED: If you want an extra challenge, forumlate this problem in matrix notation and then use CVXPY to solve.</li>
        <br>
    </ul>
</font>

<hr style="border: 20px solid black">

<h2>Bond Dedication</h2>
<br>
<font size="+1">
    <ul>
        <li>Even if you aren't interest in bonds, I claim the basics are very important to know for all aspects of business.</li>
        <br>
        <li>Why? </li>
        <br>
        <ul>
            <li>Because a bond is a fancy word for a type of loan.</li>
            <br>
            <li>Everyone interacts with loans if they have a credit card, if they have a mortgage, or if they are involved in any type of funding for a business.</li>
            <br>
            <li>Using the vocabulary of bonds, one can say that selling a bond is equivalent to taking out a loan or borrowing capital, and buying a bond is equivalent to lending capital.</li>
            <br>
        </ul>
        <hr style="border: 10px solid black">
        <br>
        <li style="color:blue">A common application of optimization for decision making within a company is to cover <b>known</b> liabilities (unknown liabilities are a different story) by constructing a <b>dedicated</b> fixed-income (bond) portfolio.</li>
        <br>
        <li style="color:blue">A <b>dedicated</b> bond portfolio is a portfolio of bonds constructed today with cash flows that offset the liabilities.</li>
        <br>
        <li style="color:blue">Often, when the liabilities span multiple years, many companies make the simplifying assumption that the only sources of risk are changes in interest rates (more specifically, the term structure of interest rates).</li>
        <br>
        <li>Consider a business facing the general problem of funding a stream of liabilities, such as labor costs, or the planning of an acquisition, expansion, product development, or other operational costs, that extends over the future.</li>
        <br>
        <li>Assume you have a forecast of future liabilities, which is reasonably accurate (not so easy to do in practice).</li>
        <br>
        <hr style="border: 10px solid black">
        <br>
        <li>Specifically, suppose a pension fund needs to cover some liabilities, due to people retiring, in the next six years.</li>
        <br>
        <li>The cash requirements (in millions) needing to be funded are:<table>
  <tr>
    <th>Year</th>
    <th>1</th>
    <th>2</th>
    <th>3</th>
      <th>4</th>
      <th>5</th>
      <th>6</th>
  </tr>
  <tr>
    <td>Liability Needing To Be Covered</td>
    <td>100</td>
    <td>200</td>
    <td>800</td>
      <td>100</td>
      <td>800</td>
      <td>1200</td>
  </tr>
            </table></li>
        <br>
        <li>Assume the pension fund can invest in the following ten government bonds of various maturities with the given cash flows and current prices (both in millions): <table>
  <tr>
    <th>Bond $\downarrow$,  Year $\rightarrow$</th>
    <th>1</th>
    <th>2</th>
    <th>3</th>
      <th>4</th>
      <th>5</th>
      <th>6</th>
      <th>Price</th>
  </tr>
  <tr>
    <td>Bond 1</td>
    <td>10</td>
    <td>10</td>
    <td>10</td>
      <td>10</td>
      <td>10</td>
      <td>100+10</td>
      <td>109</td>
  </tr>
              <tr>
    <td>Bond 2</td>
    <td>7</td>
    <td>7</td>
    <td>7</td>
      <td>7</td>
      <td>7</td>
      <td>100+7</td>
      <td>94.8</td>
  </tr>
              <tr>
    <td>Bond 3</td>
    <td>8</td>
    <td>8</td>
    <td>8</td>
      <td>8</td>
      <td>8</td>
      <td>100+8</td>
      <td>99.5</td>
  </tr>
              <tr>
    <td>Bond 4</td>
    <td>6</td>
    <td>6</td>
    <td>6</td>
      <td>6</td>
      <td>100+6</td>
      <td>n/a</td>
      <td>93.1</td>
  </tr>
              <tr>
    <td>Bond 5</td>
    <td>7</td>
    <td>7</td>
    <td>7</td>
      <td>7</td>
      <td>100+7</td>
      <td>n/a</td>
      <td>97.2</td>
  </tr>
              <tr>
    <td>Bond 6</td>
    <td>5</td>
    <td>5</td>
    <td>5</td>
      <td>100+5</td>
      <td>n/a</td>
      <td>n/a</td>
      <td>92.7</td>
  </tr>
              <tr>
    <td>Bond 7</td>
    <td>10</td>
    <td>10</td>
    <td>100+10</td>
      <td>n/a</td>
      <td>n/a</td>
      <td>n/a</td>
      <td>110</td>
  </tr>
              <tr>
    <td>Bond 8</td>
    <td>8</td>
    <td>8</td>
    <td>100+8</td>
      <td>n/a</td>
      <td>n/a</td>
      <td>n/a</td>
      <td>104</td>
  </tr>
              <tr>
    <td>Bond 9</td>
    <td>7</td>
    <td>100+7</td>
    <td>n/a</td>
      <td>n/a</td>
      <td>n/a</td>
      <td>n/a</td>
      <td>102</td>
  </tr>
              <tr>
    <td>Bond 10</td>
    <td>100</td>
    <td>n/a</td>
    <td>n/a</td>
      <td>n/a</td>
      <td>n/a</td>
      <td>n/a</td>
      <td>95.2</td>
  </tr>
            </table>
        </li>
        <br>
        <li>Find the least expensive portfolio of bonds whose cash flows will be sufficient to cover the cash requirements. This portfolio of bonds is the <b>dedicated portfolio</b>.</li>
        <br>
        <li>After every year, the dedicated bond portfolio must cover the liability for that year, but it can also pay more than the liability, which is called a surplus.</li>
        <br>
        <li>The yearly surplus is something the pension fund can choose to control.</li>
        <br>
        <li>Assume any surplus cash can be carried from one year to the next, but will earn zero interest.</li>
        <br>
        <li style="color:blue">The goal of this exercise is to formulate and solve a linear optimization problem.</li>
        <br>
        <hr style="border: 10px solid black">
        <br>
        <li style="color:red">NOT REQUIRED: How might you write these objectives and constraints in matrix notation? It isn't necessary to use a matrix formulation, however it will considerably reduce the number of things you need to type.</li>
        <br>
        <li style="color:red">NOT REQUIRED: How might you write these objectives and constraints in matrix notation using an <i>abstract</i> formulation?</li>
        <br>
        <hr style="border: 10px solid black">
        <br>
    </ul>
</font>

<h3>Decision Variables</h3>
<br>
<font size="+1">
    <ul>
        <li>What are the decision variables and how should you formulate them?</li>
        <br>
    </ul>
</font>

In [9]:
portfolio_units = cp.Variable(10, nonneg=True)

unit_1 = portfolio_units[0]
unit_2 = portfolio_units[1]
unit_3 = portfolio_units[2]
unit_4 = portfolio_units[3]
unit_5 = portfolio_units[4]
unit_6 = portfolio_units[5]
unit_7 = portfolio_units[6]
unit_8 = portfolio_units[7]
unit_9 = portfolio_units[8]
unit_10 = portfolio_units[9]


surplus = cp.Variable(6, nonneg = True)

surplus_year_1 = surplus[0]
surplus_year_2 = surplus[1]
surplus_year_3 = surplus[2]
surplus_year_4 = surplus[3]
surplus_year_5 = surplus[4]
surplus_year_6 = surplus[5]

<h3>Objective</h3>
<br>
<font size="+1">
    <ul>
        <li>What is the objective that the pension fund seeks to optimize (maximize or minimize) and how should they formulate the objective using the decision variables and the parameters about the bonds?</li>
        <br>
    </ul>
</font>

In [10]:
objective = cp.Minimize(unit_1 * 109 + unit_2 * 94.8 + unit_3 * 99.5 + unit_4 * 93.1 + unit_5 * 97.2 +
                       unit_6 * 92.7 + unit_7 * 110 + unit_8 * 104 + unit_9 * 102 + unit_10 * 95.2)

<h3>Constraints</h3>
<br>
<font size="+1">
    <ul>
        <li>What are the constraints imposed the cash flow liabilities on the pension fund, and how should they be formulated using the decision variables and the parameters about the bonds used to construct the dedicated portfolio, and how does the surplus affect these consraints?</li>
        <br>
        <ul>
            <li><i>Hint: for all years, the pension fund needs the cash flow coming from the dedicated bond portfolio to equal the liability plus some surplus;</i></li>
            <br>
            <ul>
                <li><i>and the surplus gained in a given year gets carried (with zero interest) to the cash flow coming from the dedicated bond portfolio in the subsequent year.</i></li>
            <br>
            </ul>
        </ul>
    </ul>
</font>

In [11]:
constraint_1 = [unit_1 * 10 + unit_2 * 7 + unit_3 * 8 
                + unit_4 * 6 + unit_5 * 7 + unit_6 * 5 +
unit_7 * 10 + unit_8 * 8 + unit_9 * 7 + unit_10 * 100 == 100 + surplus_year_1]


constraint_2 = [unit_1*10 + unit_2*7 + unit_3*8 
                 + unit_4*6 + unit_5*7 + unit_6*5 +unit_7*10 + unit_8*8 + unit_9*107 == 200 + surplus_year_2]

constraint_3 = [unit_1*10 + unit_2*7 + unit_3*8 + unit_4*6 + unit_5*7 + unit_6*5 +
unit_7*110 + unit_8*108 == 800 + surplus_year_3]


constraint_4 = [unit_1*10 + unit_2*7 + unit_3*8 
                + unit_4*6 + unit_5*7 + unit_6*105 == 100 + surplus_year_4]

constraint_5 = [unit_1*10 + unit_2*7 + unit_3*8 
                + unit_4*106 + unit_5*107 == 800 + surplus_year_5]


constraint_6 = [unit_1*110 + unit_2*107 + unit_3*108 == 1200 + surplus_year_6]

In [12]:
constraints = constraint_1 + constraint_2 + constraint_3 + constraint_4 + constraint_5 + constraint_6

<h3>Solution</h3>
<br>
<font size="+1">
    <ul>
        <li>What is the solution to the problem?</li>
        <br>
        <li>How can you interpret the solution?</li>
        <br>
    </ul>
</font>

In [13]:
prob = cp.Problem(objective, constraints)

In [14]:
prob.solve()

2381.1388317859164

In [15]:
np.round(portfolio_units.value, 2)

array([ 0.  , 11.21,  0.  ,  6.81,  0.  ,  0.  ,  0.  ,  6.3 ,  0.28,
        0.  ])

- This means that the least expensive portfolio is about 2381 million dollars, which contains: 11.21 units of bond 2, 6.81 units of bond 4, 6.3 units of bond 8, 0.28 units of bond 9.

<hr style="border: 20px solid black">

<h3 style="color:red">ASIDE: Abstract (General) Formulation - Linear Optimization Model for Bond Dedication</h3>

<br>
<font size="+1">
    <ul>
        <li style="color:red">NOT REQUIRED: If you want an extra challenge, forumlate this problem in matrix notation for an abstract formulation, and then use CVXPY to solve.</li>
        <br>
    </ul>
</font>

<h3 style="color:green">ASIDE: Some Practical Details About Bonds</h3>
<br>
<font size="+1" style="color:green">
    <ul>
        <li><b>YOU MAY SKIP THIS ENTIRE SUB-SECTION, IT IS NOT NECESSARY TO COMPLETE THE PROBLEM.</b></li>
        <br>
        <li>There are many particular details about the way bonds are quoted and traded in actual exchanges. This is known as bond market microstructure.</li>
        <br>
        <li>Specifically, we are only going to discuss some shallow facts regarding vanilla treasury bonds.</li>
        <br>
        <li style="color:blue"><b>Principal Value, Coupon Payments, Clean and Dirty Prices</b></li>
        <br>
        <li>Below are some definitions.</li>
        <br>
        <ul>
            <li>The <b>principal value</b> (also known as <b>par</b> or simply <b>principal</b>) of a bond is the amount that the bond issuer agrees to repay the bondholder.</li>
            <br>
            <li>The <b>term to maturity</b> of a bond is the time remaining until principal payment.</li>
            <br>
            <li>The <b>maturity date</b> of a bond is the date when the issuer will pay the principal.</li>
            <br>
            <li>The <b>coupon rate</b> or <b>nominal rate</b> of a bond is the annual interest rate that the issuer pays the bondholder.</li>
            <br>
        </ul>
        <li>Treasury bonds pay their coupons semiannually.</li>
        <br>
        <ul>
            <li>For example, a bond with an $8\%$ coupon rate and a principal of $\$1,000$ will pay a $\$40 = \$1,000 \times \frac{0.08}{2}$ installment to the holder every six months. At the maturity date, it will pay the $\$40$ installment plus the $\$1,000$ principal.</li>
            <br>
        </ul>
        <li>When an investor purchases a bond between coupon payments, the investor must compensate the seller of the bond for the coupon interest earned since the last coupon payment. This is called the <b>accrued interest</b> and is computed based on the proportion of time since the last coupon payment.</li>
        <br>
        <li>The convention for U.S. treasuries is not to include the accrued interest in the price quote. This price without the accrued interest is called the <b>clean price</b>.</li>
        <br>
        <li>It is customary to present the price quote as a percentage of the par value of the bond.</li>
        <br>
        <li>The clean price plus the accrued interest is called the <b>dirty price</b> or <b>full price</b>.</li>
        <br>
        <li>Consider an example.</li>
        <br>
        <ul>
            <li>Suppose on February $15^{th}$, 2051, investor $B$ buys a treasury bond with $\$10,000$ face value (par), $5.5\%$ coupon rate that matures on January 31, 2053.</li>
            <br>
            <li>In this case, the coupon payment is $\$275 = \frac{0.055}{2} \times \$10,000$ and the accrued interest is $\frac{15}{181}\times 275 = 22.79$.</li>
            <br>
            <li>Assume the price quote on February 15, 2051 is $101.45$. Then the full price of the bond, that is, the price paid by the buyer to the seller, is $\$ 10,114.5 + \$ 22.79 = \$ 10,137.29$.</li>
            <br>
        </ul>
        <li style="color:blue"><b>Yield Curve and Term Structure</b></li>
        <br>
        <li>Recall, the <b>yield</b> of a bond is the interest rate that makes the discounted value of the cash flows match the current price of the bond. In a way, it is the <i>implied interest rate</i> (similar to implied volatility being the volatility that makes the Black-Scholes pricing formula match the current price of the option).</li>
        <br>
        <li>By convention, the yield is quoted on an annualized basis.</li>
        <br>
        <li>The <b>treasury yield curve</b> is the curve of yields for <b>on-the-run</b> (most recently auctioned, i.e hot-off-the-press) treasuries.</li>
        <br>
        <li>The <b>term structure of interest rates</b> is the curve of yields on a risk-free <b>zero-coupon</b> bond as a function of maturity. Note that zero-coupon bonds do not actually trade, except the edge case of <a href="https://www.treasurydirect.gov/instit/marketables/strips/strips.htm">Treasury STRIPS</a>.</li>
        <br>
        <li>Note: Depending on how you precisely define the yield curve and term structure of rates, it can be the case that the yield curve $\neq$ term structure of interest rates.</li>
        <br>
        <ul>
            <li>This is because treasuries with maturity greater than one year are not zero-coupon bonds.</li>
            <br>
            <li>Specifically, the term structure of interest rates is actually a theoretical construct that must be estimated from actual traded bonds.</li>
            <br>
            <li>This begs the question of how to estimate the term structure of interest rates?</li>
            <br>
        </ul>
        <li>A quick and dirty approach is to completely ignore the subtle difference between term structure of rates and yield curve and use the yield curve as a proxy for the term structure of interest rates.</li>
        <br>
        <li>Another approach is to use a <b>bootstrapping</b> approach.</li>
        <br>
        <ul>
            <li>Use several coupon bonds with various maturities.</li>
            <br>
            <li>Determine the spot rate (present value) implied by the bond with the shortest maturity.</li>
            <br>
            <li>Use that knowledge to compute the spot rate implied by the bond with the next shortest maturity, and so on.</li>
            <br>
        </ul>
        <li>Another more elaborate approach to estimating the term structure is to consider all bonds with similar characteristics available in the market and perform a regression analysis. You can learn more about this approach in <a href="https://www.google.com/books/edition/The_Econometrics_of_Financial_Markets/7Gkri6HWWkgC?hl=en&gbpv=1&printsec=frontcover">The Econometrics of Financial Markets</a>.</li>
        <li>If you want to read more, see <a href="https://www.pimco.com/en-us/resources/education/everything-you-need-to-know-about-bonds/">PIMCO's notes on everything you need to know about bonds.</a></li>
        <br>
        <li>If you want to read more about corporate or municipal bond markets, see <a href="https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2661801">Transaction Costs, Trade Throughs, and Riskless Principal Trading in Corporate Bond Markets</a> and <a href="https://www.tandfonline.com/doi/abs/10.2469/faj.v71.n6.3">The Structure of Trading in Bond Markets</a></li>
        <br>
    </ul>
</font>

<hr style="border: 20px solid black">

<h2>Quadratic Asset Allocation</h2>
<br>
<font size="+1">
    <ul>
        <li>Assume you have estimated, with the best possible machine learning algorithm, the one-year correlations and standard deviations of the following groups of large stocks, small stocks, and bonds.</li>
        <br>
        <li><table>
  <tr>
    <th bgcolor="dodgerblue">Correlations</th>
    <th>Large Stocks</th>
    <th>Small Stocks</th>
    <th>Bonds</th>
      <th bgcolor="pink">Standard Deviation</th>
  </tr>
  <tr>
    <td>Large Stocks</td>
    <td bgcolor="dodgerblue" >1</td>
    <td bgcolor="dodgerblue">0.6</td>
    <td bgcolor="dodgerblue">0.2</td>
      <td bgcolor="pink">0.12</td>
  </tr>
  <tr>
    <td>Small Stocks</td>
    <td bgcolor="dodgerblue">0.6</td>
    <td bgcolor="dodgerblue">1</td>
    <td bgcolor="dodgerblue">0.5</td>
      <td bgcolor="pink">0.2</td>
  </tr>
  <tr>
    <td>Bonds</td>
    <td bgcolor="dodgerblue">0.2</td>
    <td bgcolor="dodgerblue">0.5</td>
    <td bgcolor="dodgerblue">1</td>
      <td bgcolor="pink">0.05</td>
  </tr>
</table></li>
        <br>
        <li>Recall the covariance of asset $i$ and asset $j$ equals the product of the correlation of asset $i$ and $j$ with the standard deviation of asset $i$ and the standard deviation of asset $j$.</li>
        <br>
        <li>After doing some computations, you compute the covariances corresponding to the different asset classes are as follows: </li>
        <br>
        <li><table>
  <tr>
    <th bgcolor="orange">Covariance</th>
    <th>Large Stocks</th>
    <th>Small Stocks</th>
    <th>Bonds</th>
  </tr>
  <tr>
    <td>Large Stocks</td>
    <td bgcolor="orange" >0.0144</td>
    <td bgcolor="orange">0.0144</td>
    <td bgcolor="orange">0.0012</td>
  </tr>
  <tr>
    <td>Small Stocks</td>
    <td bgcolor="orange">0.0144</td>
    <td bgcolor="orange">0.04</td>
    <td bgcolor="orange">0.005</td>
  </tr>
     <tr>
    <td>Bonds</td>
    <td bgcolor="orange">0.0012</td>
    <td bgcolor="orange">0.005</td>
    <td bgcolor="orange">0.0025</td>
  </tr>
</table></li>
        <br>
        <li>Given the above estimates, determine the asset allocation that produces the smallest amount of risk.</li>
        <br>
        <li>Specifically, find a portfolio comprised of these three asset classes whose return has the lowest variance.</li>
        <br>
        <li>Assume the portfolio can only hold long positions in each of the three asset classes and you are to be fully invested.</li>
        <br>
        <li style="color:blue">The goal of this exercise is to formulate and solve a quadratic (convex) optimization problem.</li>
        <br>
        <li style="color:blue">CVXPY requires you to write these objectives and constraints in matrix notation.</li>
        <br>
        <li style="color:blue">In order to formulate this problem, you need to do the following.</li>
        <br>
        <ul style="color:blue">
            <li>First, construct the <b>covariance matrix</b>.</li>
            <br>
            <li>Next, to construct the objective, you will have to use the <b>cp.quad_form(w, V)</b> method from CVXPY.</li>
            <br>
            <li>Finally, the problem can be solved <i>"as usual"</i>.</li>
            <br>
        </ul>
        <hr style="border: 10px solid black">
        <br>
        <li style="color:red">NOT REQUIRED: How might you write these objectives and constraints in matrix notation using an <i>abstract</i> formulation?</li>
        <br>
        <hr style="border: 10px solid black">
        <br>
    </ul>
</font>

<h3>Decision Variables</h3>
<br>
<font size="+1">
    <ul>
        <li>What are the decision variables and how should you formulate them?</li>
        <br>
    </ul>
</font>

In [16]:
std_correlations = pd.DataFrame({'Large Stocks': [1, 0.6, 0.2],
                              'Small Stocks': [0.6, 1, 0.5],
                                 'Bonds': [0.2, 0.5, 1],
                                'Standard Deviation': [0.12, 0.2, 0.5]}, index = ['Large Stocks',  'Small Stocks', 'Bonds'])

In [17]:
std_correlations

Large Stocks  Small Stocks  Bonds  Standard Deviation
Large Stocks           1.0           0.6    0.2                0.12
Small Stocks           0.6           1.0    0.5                0.20
Bonds                  0.2           0.5    1.0                0.50

In [18]:
covariance_matrix = pd.DataFrame({'Large Stocks': [0.0144, 0.0144, 0.0012],
                              'Small Stocks': [0.0144, 0.04, 0.005],
                                 'Bonds': [0.0012, 0.005, 0.0025]}, index = ['Large Stocks',  'Small Stocks', 'Bonds'])

In [19]:
covariance_matrix

Large Stocks  Small Stocks   Bonds
Large Stocks        0.0144        0.0144  0.0012
Small Stocks        0.0144        0.0400  0.0050
Bonds               0.0012        0.0050  0.0025

In [20]:
# Decision variables

portfolio_weights = cp.Variable(len(std_correlations))

weight_Large_Stock = portfolio_weights[0]

weight_Small_Stock = portfolio_weights[1]

weight_Bond = portfolio_weights[2]

<h3>Objective</h3>
<br>
<font size="+1">
    <ul>
        <li>What is your objective that you seek to optimize (maximize or minimize) and how should you formulate the objective using the decision variables and the parameters about the asset classes?</li>
        <br>
    </ul>
</font>

In [21]:
objective = cp.Minimize(cp.quad_form(portfolio_weights, covariance_matrix))

<h3>Constraints</h3>
<br>
<font size="+1">
    <ul>
        <li>What are your constraints, and how should they be formulated using the decision variables and the parameters about the asset classes?</li>
        <br>
    </ul>
</font>

In [22]:
constraint_no_short_selling = [weight_Large_Stock >= 0,
                              weight_Small_Stock >= 0,
                              weight_Bond >= 0]

constraint_no_leverage = [weight_Large_Stock <= 1,
                          weight_Small_Stock <= 1,
                          weight_Bond <= 1]

constraint_fully_vested = [weight_Large_Stock + weight_Small_Stock + weight_Bond == 1]

constraints = constraint_no_short_selling + constraint_no_leverage + constraint_fully_vested

<h3>Solution</h3>
<br>
<font size="+1">
    <ul>
        <li>What is the solution to the problem?</li>
        <br>
        <li>How can you interpret the solution?</li>
        <br>
    </ul>
</font>

In [23]:
prob = cp.Problem(objective, constraints)

In [24]:
prob.solve()

0.002383448275862069

In [25]:
portfolio_weights.value

array([8.96551724e-02, 9.29600321e-22, 9.10344828e-01])

In [26]:
try:
    prob.solve()
    print ("Optimal portfolio")
    print ("----------------------")
    for s in range(len(std_correlations)):
        print (" Investment in {} : {}% of the portfolio".format(std_correlations.index[s],
                                                                 round(100*portfolio_weights.value[s],2)))
    print ("----------------------")
    print ("Expected risk    = {}%".format(round(100*((cp.quad_form(portfolio_weights.value, covariance_matrix).value)**(1/2)),2)))
except:
    print ("Error")

Optimal portfolio
----------------------
 Investment in Large Stocks : 8.97% of the portfolio
 Investment in Small Stocks : 0.0% of the portfolio
 Investment in Bonds : 91.03% of the portfolio
----------------------
Expected risk    = 4.88%


<hr style="border: 20px solid black">